pilot/pkg/proxy/envoy/v2/datastructures.go

```diff
@@ -0,0 +1,139 @@
+// Copyright 2018 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+
+package v2
+
+import (
+	"sync"
+
+	xdsapi "github.com/envoyproxy/go-control-plane/api"
+	"istio.io/istio/pilot/pkg/model"
+)
+
+// Endpoints (react to updates in registry endpoints) (notify clusters)
+// 1. Endpoints can be shared across multiple clusters.
+// 2. The membership of an endpoint in a cluster is based on the set of labels
+//    associated with the endpoint and the set of label selectors associated with the cluster.
+// 3. When a new endpoint is added, the label selectors associated with the clusters have
```

Comment:
How ? Linear search in the clusters first, and later more advanced if needed ?

Also: once it's added, do we push anything to the cluster, or wait for next EDS call ?